src/distributions/weighted_index.rs

```diff
@@ -130,6 +131,66 @@ impl<X: SampleUniform + PartialOrd> WeightedIndex<X> {
         })
     }
 
+    /// Reuses the weighted index by assigning a new set of weights. This can change the number
+    /// of weights if the amount contained in `weights` is different from what was used to
+    /// initialize the index. Can cause a reallocation if the iterator contains more weights.
+    ///
+    /// ***Panics*** if the index is reused after encountering an error.
+    pub fn assign_new_weights<I>(&mut self, weights: I) -> Result<(), WeightedError >
+    where
+        I: IntoIterator,
+        I::IntoIter: ExactSizeIterator,
+        I::Item: SampleBorrow<X>,
+        X: for<'a> ::core::ops::AddAssign<&'a X> + Clone + Default,
+    {
+        match self.inner_assign_new_weights(weights)
+        {
+            Ok(()) => Ok(()),
+            e @ Err(_) => {
+                self.cumulative_weights.clear();
+                e
+            }
+        }
+    }
+
+    fn inner_assign_new_weights<I>(&mut self, weights: I) -> Result<(), WeightedError >
+    where
+        I: IntoIterator,
+        I::IntoIter: ExactSizeIterator,
+        I::Item: SampleBorrow<X>,
+        X: for<'a> ::core::ops::AddAssign<&'a X> + Clone + Default,
+    {
+        let mut iter = weights.into_iter();
+
+        if iter.len() != self.cumulative_weights.len() + 1 {
+            return Err(WeightedError::LenMismatch);
+        }
+
+        let mut total_weight: X = iter.next().ok_or(WeightedError::NoItem)?.borrow().clone();
+
+        let zero = <X as Default>::default();
+        if !(total_weight >= zero) {
+            return Err(WeightedError::InvalidWeight);
+        }
+        let mut contains_invalid = false;
+        for (w, c) in iter.zip(self.cumulative_weights.iter_mut()) {
+            contains_invalid |= !(w.borrow() >= &zero);
+            *c = total_weight.clone();
+            total_weight += w.borrow();
+        }
+        if contains_invalid {
+            return Err(WeightedError::InvalidWeight);
+        } else if total_weight == zero {
+            return Err(WeightedError::AllWeightsZero);
+        };
+
+        self.weight_distribution = X::Sampler::new(zero, total_weight.clone());
```

Comment:
There's still a problem: this panics if `total_weight` is +inf, and we don't catch panics.